server/src/main/java/org/elasticsearch/action/bulk/BulkResponse.java

```diff
@@ -49,28 +50,42 @@
     private static final String ERRORS = "errors";
     private static final String TOOK = "took";
     private static final String INGEST_TOOK = "ingest_took";
+    private static final String ITEMS_OMITTED = "items_omitted";
 
     public static final long NO_INGEST_TOOK = -1L;
 
     private final BulkItemResponse[] responses;
     private final long tookInMillis;
     private final long ingestTookInMillis;
+    @Nullable
+    private final Boolean noItemsOnSuccess;
 
     public BulkResponse(StreamInput in) throws IOException {
         super(in);
         responses = in.readArray(BulkItemResponse::new, BulkItemResponse[]::new);
         tookInMillis = in.readVLong();
         ingestTookInMillis = in.readZLong();
+        noItemsOnSuccess = in.readOptionalBoolean();
     }
 
     public BulkResponse(BulkItemResponse[] responses, long tookInMillis) {
         this(responses, tookInMillis, NO_INGEST_TOOK);
     }
 
     public BulkResponse(BulkItemResponse[] responses, long tookInMillis, long ingestTookInMillis) {
-        this.responses = responses;
+        this(responses, tookInMillis, ingestTookInMillis, false);
+    }
+    public BulkResponse(BulkItemResponse[] responses, long tookInMillis, Boolean noItemsOnSuccess) {
+        this(responses, tookInMillis, NO_INGEST_TOOK, noItemsOnSuccess);
+    }
+
+    public BulkResponse(BulkItemResponse[] responses, long tookInMillis, long ingestTookInMillis, Boolean noItemsOnSuccess) {
+        final boolean hasError = responses != null && Arrays.stream(responses).anyMatch(r-> r != null && r.isFailed());
+        this.responses = noItemsOnSuccess != null && noItemsOnSuccess && hasError == false
```

Comment:
maybe add a comment that the bulk items are omitted because these are no longer needed if `noItemsOnSuccess ` is true, which would allow the gc to clean the items up before the response has been serialized over the network.